### PR TITLE
Improve SchemaManager logic for comparing text indexes

### PR DESCRIPTION
**Pull request by jmikola**
|      Q       |   A
|------------- | -----------
| Type         | bug
| BC Break     | no
| Fixed issues | #1821

#### Summary

This is a cherry-pick of https://github.com/doctrine/mongodb-odm/pull/1828 for `master` and includes various CS fixes. I also added two tests for the loose equality checks for keys/weights.

### CODE REVIEW

File: lib/Doctrine/ODM/MongoDB/SchemaManager.php [hunk 1]
```diff
@@ -428,9 +429,89 @@ public function isMongoIndexEquivalentToDocumentIndex($mongoIndex, $documentInde
...
+            unset($mongoIndexKeys['_fts'], $mongoIndexKeys['_ftsx']);
+
+            $documentIndexKeys = array_filter($documentIndexKeys, function ($type) {
+                return $type !== 'text';
+            });
+        }
+
+        /* Avoid a strict equality check here. The numeric type returned by
+         * MongoDB may differ from the document index without implying that the
+         * indexes themselves are inequivalent. */
```
Comment (jmikola) [inline lib/Doctrine/ODM/MongoDB/SchemaManager.php]: See https://github.com/doctrine/mongodb-odm/pull/1828#discussion_r203394981 for additional context.

File: lib/Doctrine/ODM/MongoDB/SchemaManager.php [hunk 2]
```diff
@@ -428,9 +429,89 @@ public function isMongoIndexEquivalentToDocumentIndex($mongoIndex, $documentInde
...
         }
 
+        if (isset($mongoIndex['weights']) && ! $this->isEquivalentTextIndexWeights($mongoIndex, $documentIndex)) {
+            return false;
+        }
+
+        foreach (['default_language', 'language_override', 'textIndexVersion'] as $option) {
+            /* Text indexes will always report defaults for these options, so
+             * only compare if we have explicit values in the document index. */
+            if (isset($mongoIndex[$option]) && isset($documentIndexOptions[$option]) &&
```
Comment (malarzm) [inline lib/Doctrine/ODM/MongoDB/SchemaManager.php]: `isset`s can be combined to one `isset($mongoIndex[$option], $documentIndexOptions[$option])`

File: lib/Doctrine/ODM/MongoDB/SchemaManager.php [hunk 3]
```diff
@@ -428,9 +429,89 @@ public function isMongoIndexEquivalentToDocumentIndex($mongoIndex, $documentInde
...
+        foreach ($documentIndex['keys'] as $key => $type) {
+            if ($type !== 'text' || isset($documentIndexWeights[$key])) {
+                continue;
+            }
+
+            $documentIndexWeights[$key] = 1;
+        }
+
+        /* MongoDB returns the weights sorted by field name, but we'll sort both
+         * arrays in case that is internal behavior not be be relied upon. */
```
Comment (malarzm) [inline lib/Doctrine/ODM/MongoDB/SchemaManager.php]: doubled "be" :)

Comment (jmikola) [reply lib/Doctrine/ODM/MongoDB/SchemaManager.php]: Ah, I meant to write "to be". Good catch!

File: lib/Doctrine/ODM/MongoDB/SchemaManager.php (hunk 2 above)
Comment (jmikola) [reply lib/Doctrine/ODM/MongoDB/SchemaManager.php]: I had copied this from the `['bits', 'max', 'min']` loop above. Will make the change in both places.

File: lib/Doctrine/ODM/MongoDB/SchemaManager.php (hunk 3 above)
Comment (jmikola) [reply lib/Doctrine/ODM/MongoDB/SchemaManager.php]: Also fixed in 1.2.x: https://github.com/doctrine/mongodb-odm/commit/70441acd0f2809445f1ae7a61acbbbd3ee117b42

**Comment by jmikola** [comment]
Please ignore the `continuous-integration/travis-ci/push` build. I inadvertently pushed the branch to the `doctrine` repository instead of my fork.

**Pull request merged**